Upgrading built-in Python from Python 3.12 to Python 3.14

roddyrap opened the pull request:
Closes #39

roddyrap committed:
Upgrading static python to python3.14

roddyrap committed:
Added custom autoconf installer

This is helpful to manually run in the docker and then regenerate
configure files for the different submodules.

It's not used automatically.

roddyrap committed:
Re-added modules needed for gef support

I assume some of the depended modules' internal imports changed.

guyush1 commented on Dockerfile:
```diff
@@ -24,19 +24,22 @@ RUN apt update && apt install -y \
     pkg-config \
-    python3.12 \
     python3-pip \
+    software-properties-common \
```
Could you keep the list alphabetically ordered (also need to move libtool as well haha)

roddyrap replied:
sure

guyush1 commented on src/compilation/build.sh:
```diff
@@ -482,13 +485,14 @@ function build_python() {
 
     # Having a python-config file is an indication that we successfully built python.
-    if [[ -f "$python_lib_dir/python-config" ]]; then
+    if [[ -f "$python_lib_dir/python-config" && -f "$python_lib_dir/libpython3.14.a" ]]; then
```
try to make it generic by fetching the python version via python --version or which pythonX (might need to create a symlink so that it's name won't change each time we change a version, but you get the gist)

guyush1 replied:
EDIT - even a global variable exported in the dockerfile is more then fine by me :)

guyush1 commented on src/compilation/build.sh:
```diff
@@ -499,14 +503,14 @@ function build_python() {
     export MODULE_BUILDTYPE="static"
     export CONFIG_SITE="$python_dir/config.site-static"
-    >&2 CFLAGS="-static" LDFLAGS="-static" ../configure \
+    >&2 CFLAGS="${CFLAGS} -static" LDFLAGS="${LDFLAGS} -static -llzma" ../configure \
```
LOL good catch

guyush1 commented on src/compilation/build.sh:
```diff
@@ -499,14 +503,14 @@ function build_python() {
         --build=x86_64-pc-linux-gnu \
         --host=$HOST \
-        --with-build-python=/usr/bin/python3.12 \
+        --with-build-python=/usr/bin/python3.14 \
```
same comment as before about making it generic

guyush1 commented on src/compilation/build.sh:
```diff
@@ -517,7 +521,7 @@ function build_python() {
     # Regenerate frozen modules with gdb env varaible. Do it after the configure because we need
     # the `regen-frozen` makefile.
-    >&2 python3.12 ../Tools/build/freeze_modules.py
+    >&2 python3.14 ../Tools/build/freeze_modules.py
```
same

guyush1 commented on src/compilation/install_autoconf.sh:
I do not see anyone using this specific script.
I remember we talked (quiet a while ago) about not having the latest autoconf available giving you issues with issue #63 
Is this script related to that?

Anyhow, please document the purpose of this script :)

roddyrap replied:
It's not used automatically anywhere right now but I used it manually when I modify the configure.ac in cpython.

I literally don't have the correct autoconf versions in my PC's package manager or in the docker, so I have to install it manually. At this point, might as well share how I do it with the rest of the world.

guyush1 commented on src/compilation/install_autoconf.sh:
```diff
@@ -0,0 +1,33 @@
+
+# Example: "2.10"
+AUTOCONF_VERSION="${1}"
+if [ -z "${AUTOCONF_VERSION}" ]; then
```
```suggestion
if [[ -z "${AUTOCONF_VERSION}" ]]; then
```

roddyrap replied:
Why though?

guyush1 replied:
[ is an executable, [[ is a builtin. No reason to use executables when a builtin does the trick.

guyush1 commented on src/compilation/frozen_python_modules.txt:
```diff
@@ -1,4 +1,8 @@
 _aix_support
+_colorize
```
could you tag the needed modules for GEF via a comment - i don't want them getting removed accidentally in the future (thank you for not remembering it :) )

roddyrap replied:
Not really, this file doesn't support comments...
I can add comment support but I really don't want to do it right now....

guyush1 commented on src/submodule_packages/cpython-static:
Looking good 💯

guyush1 replied:
EDIT: btw, since you added expat, we can add ncurses as well, no? we already support it in gdb static just like expat

roddyrap replied:
Yeah, I think? I think it's better to do in a separate issue though so I'll open one.

roddyrap replied:
#82

guyush1 reviewed (changes requested):
Awesome! looking forward to merge this, just have a look over my comments :P

guyush1 commented:
@roddyrap when trying to use GEF.py with an ARM build, i got the following error:
`Error occurred in Python: No module named '_ctypes'`
does not replicate in an x86 build.

In fact, here is an easier replication of the bug:
```
gdb> python
import _ctypes
CTRL+D
```

There should be an error in a non-x86 build (or at the very least in arm).

could you look at that?

guyush1 commented:
NOTE: this MR closes #39

roddyrap commented:
> @roddyrap when trying to use GEF.py with an ARM build, i got the following error:
> `Error occurred in Python: No module named '_ctypes'`
> does not replicate in an x86 build.
> 
> In fact, here is an easier replication of the bug:
> ```
> gdb> python
> import _ctypes
> CTRL+D
> ```
> 
> There should be an error in a non-x86 build (or at the very least in arm).
> 
> could you look at that?

Ah yikes, yeah I'll check it out

guyush1 commented:
@roddyrap any updates on this?

roddyrap commented:
> > @roddyrap when trying to use GEF.py with an ARM build, i got the following error:
> > `Error occurred in Python: No module named '_ctypes'`
> > does not replicate in an x86 build.
> > In fact, here is an easier replication of the bug:
> > ```
> > gdb> python
> > import _ctypes
> > CTRL+D
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > There should be an error in a non-x86 build (or at the very least in arm).
> > could you look at that?
> 
> Ah yikes, yeah I'll check it out

The issue is that I am dumb, by the way. Libffi compilation failed lol

roddyrap committed:
Fixed libffi cross-compilation

I was dumb and forgot to specify the cross-compilation arguments for
libffi, so it only compiled for x86_64 (Or the local architecture)

This commit fixes the problem, adding the correct arguments. Also
removed hardcoded requirement to compile libffi with an x86_64 host.

roddyrap committed:
Made compiled Python version a Docker argument

This allows us to relatively easily change the python version that's
being used to build cpython. We only need to change it in the Docker,
and the build.sh script changes automatically.
